lock-api/src/main/java/com/palantir/lock/client/MultiClientTransactionStarter.java

```diff
@@ -0,0 +1,323 @@
+            UUID requestorId,
+            List<BatchElement<NamespaceAndRequestParams, List<StartIdentifiedAtlasDbTransactionResponse>>> batch) {
+
+        Map<Namespace, ResponseHandler> namespaceWiseResponseHandler = getResponseHandlers(batch);
+        MultiClientRequestManager multiClientRequestManager =
+                new MultiClientRequestManager(getNamespaceWiseRequestParams(batch));
+
+        Map<Namespace, List<StartIdentifiedAtlasDbTransactionResponse>> startTransactionResponses = new HashMap<>();
+
+        try {
+            while (multiClientRequestManager.requestsPending()) {
+                startTransactionResponses =
+                        getStartTransactionResponses(multiClientRequestManager.requestMap, delegate, requestorId);
+                startTransactionResponses.forEach((namespace, responseList) -> {
+                    multiClientRequestManager.updatePendingStartTransactionsCount(namespace, responseList.size());
+                    namespaceWiseResponseHandler.get(namespace).processResponse(responseList);
+                });
+            }
+        } finally {
+            clearResources(multiClientRequestManager, namespaceWiseResponseHandler);
+        }
+    }
+
+    private static void clearResources(
+            MultiClientRequestManager multiClientRequestManager,
+            Map<Namespace, ResponseHandler> namespaceWiseResponseHandler) {
+        multiClientRequestManager.close();
+        namespaceWiseResponseHandler.forEach((_unused, responseHandler) -> responseHandler.close());
+    }
+
+    private static Map<Namespace, ResponseHandler> getResponseHandlers(
+            List<BatchElement<NamespaceAndRequestParams, List<StartIdentifiedAtlasDbTransactionResponse>>> batch) {
+        Map<Namespace, ResponseHandler> responseHandlers = new HashMap<>();
+
+        for (BatchElement<NamespaceAndRequestParams, List<StartIdentifiedAtlasDbTransactionResponse>> element : batch) {
+            NamespaceAndRequestParams requestParams = element.argument();
+            Namespace namespace = requestParams.namespace();
+            responseHandlers
+                    .computeIfAbsent(
+                            namespace,
+                            _unused ->
+                                    new ResponseHandler(requestParams.params().lockCleanupService()))
+                    .addPendingFuture(SettableResponse.of(requestParams.params().numTransactions(), element.result()));
+        }
+
+        return responseHandlers;
+    }
+
+    @VisibleForTesting
+    static Map<Namespace, RequestParams> getNamespaceWiseRequestParams(
+            List<BatchElement<NamespaceAndRequestParams, List<StartIdentifiedAtlasDbTransactionResponse>>> batch) {
+        return batch.stream()
+                .map(BatchElement::argument)
+                .collect(Collectors.toMap(
+                        NamespaceAndRequestParams::namespace, NamespaceAndRequestParams::params, RequestParams::merge));
```

Comment:
non-actionable: I would probably spend too much time figuring out how to do this with groupingBy and not creating intermediate objects, but this is fine :)